util/linuxfw/iptables.go

```diff
@@ -823,3 +825,479 @@ type markTarget struct {
 	Mode markMode
 	Mark uint32
 }
+
+type iptablesInterface interface {
+	// Adding this interface for testing purposes so we can mock out
+	// the iptables library, in reality this is a wrapper to *iptables.IPTables.
+	Insert(table, chain string, pos int, args ...string) error
+	Append(table, chain string, args ...string) error
+	Exists(table, chain string, args ...string) (bool, error)
+	Delete(table, chain string, args ...string) error
+	ClearChain(table, chain string) error
+	NewChain(table, chain string) error
+	DeleteChain(table, chain string) error
+}
+
+type iptableRunner struct {
+	ipt4 iptablesInterface
+	ipt6 iptablesInterface
+
+	v6Available    bool
+	v6NATAvailable bool
+}
+
+func NewIptableRunner(logf logger.Logf) (*iptableRunner, error) {
+	ipt4, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	if err != nil {
+		return nil, err
+	}
+
+	v6err := checkIPv6(logf)
+	if v6err != nil {
+		logf("disabling tunneled IPv6 due to system IPv6 config: %v", v6err)
+	}
+	supportsV6 := v6err == nil
+	supportsV6NAT := supportsV6 && supportsV6NAT()
+	if supportsV6 {
+		logf("v6nat = %v", supportsV6NAT)
+	}
+
+	var ipt6 *iptables.IPTables
+	if supportsV6 {
+		ipt6, err = iptables.NewWithProtocol(iptables.ProtocolIPv6)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return newIptableRunnerAdvanced(ipt4, ipt6, supportsV6, supportsV6NAT), nil
+}
+
+// newIptableRunnerAdvanced is a helper function for testing purposes
+func newIptableRunnerAdvanced(ipt4, ipt6 iptablesInterface, supV6, supV6NAT bool) *iptableRunner {
+	return &iptableRunner{ipt4, ipt6, supV6, supV6NAT}
+}
+
+func (i *iptableRunner) GetV6Available() bool {
+	return i.v6Available
+}
+
+func (i *iptableRunner) GetV6NATAvailable() bool {
+	return i.v6NATAvailable
+}
+
+// AddLoopbackRule adds a iptables rule to permit loopback traffic to
+// a local Tailscale IP.
+func (i *iptableRunner) AddLoopbackRule(addr netip.Addr) error {
+	nf := i.ipt4
+	if addr.Is6() {
+		nf = i.ipt6
+	}
+
+	if err := nf.Insert("filter", "ts-input", 1, "-i", "lo", "-s", addr.String(), "-j", "ACCEPT"); err != nil {
+		return fmt.Errorf("adding loopback allow rule for %q: %w", addr, err)
+	}
+
+	return nil
+}
+
+// tsChain returns the name of the tailscale sub-chain corresponding
+// to the given "parent" chain (e.g. INPUT, FORWARD, ...).
+func tsChain(chain string) string {
+	return "ts-" + strings.ToLower(chain)
+}
+
+// getIptablebyProto returns the iptablesInterface in iptables runner for the given protocol.
```

Comment:
Nit: Maybe more functional? i.e.: `getIptablebyProto returns an object you can use to manipulate IP tables for the given protocol.`